private/buf/bufstudioagent/bufstudioagent_test.go

```diff
@@ -193,14 +194,31 @@ func testPlainPostHandlerErrors(t *testing.T, upstreamServer *httptest.Server) {
 	})
 
 	t.Run("invalid_upstream", func(t *testing.T) {
-		listener, err := net.Listen("tcp", "127.0.0.1:")
+		tcpAddr, err := net.ResolveTCPAddr("tcp", "127.0.0.1:")
 		require.NoError(t, err)
+		listener, err := net.ListenTCP("tcp", tcpAddr)
+		require.NoError(t, err)
+		done := make(chan struct{})
+		t.Cleanup(func() {
+			done <- struct{}{}
+		})
 		go func() {
-			conn, err := listener.Accept()
-			require.NoError(t, err)
-			require.NoError(t, conn.Close())
+			for {
+				select {
+				case <-done:
+					listener.Close()
+					return
+				default:
+					err := listener.SetDeadline(time.Now().Add(time.Second))
```

Comment:
`time.Second` is an arbitrary number, but the nice thing about it is that when you run `go test -count=1 -timeout=20s -race github.com/bufbuild/buf/private/buf/bufstudioagent` you can estimate how many connections did it accept before accepting the one coming from the studio agent.

Reply:
I'm okay with leaving this deadline for now, but for testing, I think it's fine to retry pretty aggressively (e.g. `time.Milisecond`), since we want the test to run as quickly as possible.

Reply:
Updated to 10ms: [dde47d6](https://github.com/bufbuild/buf/pull/3113/commits/dde47d68dced33e354b66dcbb698f4cf2f0f22d0), I'm not sure how what would happen with 1ms when GitHub action runners' performance is degraded.